### PR TITLE
add button and gradient border

### PR DESCRIPTION
**Pull request by upacyxou**
- gradient border added
- button from `sc` added

### CODE REVIEW

File: src/components/ui/Button.tsx [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    transitionDuration('duration-100'),
+    cursor({ 'cursor-not-allowed': !available }),
+    outlineStyle('focus:outline-none'),
+    opacity({ 'opacity-50': !available }),
+    boxShadow({
+      'shadow-2xl': available,
+      'hover:shadow-lg': available,
+      'active:shadow-button-active': available && !gradientFont,
+    }),
+    width(fullWidth ? 'w-full' : 'w-fit'),
```
Comment (T-Damer) [inline src/components/ui/Button.tsx]: `w-fit` usually works as default: 
image
So you can do `width({'w-full': fullWidth})`

File: src/components/ui/Button.tsx [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    textAlign(center ? 'text-center' : undefined),
```
Comment (T-Damer) [inline src/components/ui/Button.tsx]: Let's do an object here

File: src/components/ui/Button.tsx [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+    fontSize(small ? 'text-sm' : 'text-lg'),
+    lineHeight(small ? 'leading-5' : 'leading-7'),
+    type !== 'tertiary'
```
Comment (T-Damer) [inline src/components/ui/Button.tsx:74]: Put this inside `padding`

File: src/components/ui/Button.tsx [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+  fullWidth?: boolean,
+  center?: boolean,
+  available?: boolean,
+  small?: boolean,
+  gradientFont?: boolean
+) =>
+  classnames(
+    display('flex'),
+    flexDirection('flex-row'),
+    justifyContent(center ? 'justify-center' : undefined),
```
Comment (T-Damer) [inline src/components/ui/Button.tsx]: Let's do an object here

File: src/components/ui/Button.tsx [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+  transitionProperty,
+  transitionTimingFunction,
+  width,
+} from '@/classnames/tailwind'
+import Arrow from '@/icons/Arrow'
+import Loading from '@/icons/Loading'
+import React from 'react'
+
+const commonClasses = (
+  type: ButtonType,
```
Comment (T-Damer) [inline src/components/ui/Button.tsx]: Those args are pretty big, let's use object-prop

File: src/components/ui/Button.tsx [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+const colorClasses = (
+  type: ButtonType,
+  available?: boolean,
+  gradientFont?: boolean
+) => {
+  const hasNoGradient = !gradientFont
+
+  return classnames(
+    type === 'primary'
```
Comment (T-Damer) [inline src/components/ui/Button.tsx]: Can we use `switch-case`?

File: src/components/ui/Button.tsx [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+      'text-transparent': true,
+      'active:text-accent': available,
+    }),
+    backgroundClip('bg-clip-text'),
+    backgroundImage('bg-gradient-to-r'),
+    gradientColorStops('from-secondary', 'to-accent')
+  )
+
+interface ButtonProps {
+  fullWidth?: boolean
```
Comment (T-Damer) [inline src/components/ui/Button.tsx]: Put optional arguments to the bottom, please

File: src/icons/Loading.tsx [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  width,
+} from '@/classnames/tailwind'
+
+const icon = (small?: boolean, screenCentre?: boolean) =>
+  classnames(
+    animation('animate-spin'),
+    textColor('text-inherit'),
+    width(small ? 'w-3' : 'w-5'),
+    height(small ? 'h-3' : 'h-5'),
+    inset(screenCentre ? { 'top-1/2': true, 'left-1/2': true } : undefined),
```
Comment (T-Damer) [inline src/icons/Loading.tsx]: Let's do:
```TS
{ 'top-1/2': screenCentre, 'left-1/2': screenCentre}
```

File: src/icons/Loading.tsx [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    position(screenCentre ? 'absolute' : undefined)
```
Comment (T-Damer) [inline src/icons/Loading.tsx]: And here:
```TS
position({'absolute': screenCenter})
```

File: src/icons/Loading.tsx [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+  )
+const iconPath = classnames(opacity('opacity-100'))
```
Comment (T-Damer) [inline src/icons/Loading.tsx]: Remove the `classnames` here

File: src/icons/Loading.tsx [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+export default function ({ small, screenCentre }: LoadingProps) {
+  return (
+    <svg className={icon(small, screenCentre)} viewBox="0 0 24 24">
+      <path
+        className={iconPath}
+        fill="currentColor"
+        d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
+      ></path>
```
Comment (T-Damer) [inline src/icons/Loading.tsx]: I think we can close the path like this
```TS
<path props... />
```

File: src/components/ui/Button.tsx [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+const colorClasses = (
+  type: ButtonType,
+  available?: boolean,
+  gradientFont?: boolean
+) => {
+  const hasNoGradient = !gradientFont
+
+  switch (type) {
+    case 'primary':
+      return classnames(
```
Comment (T-Damer) [inline src/components/ui/Button.tsx:106]: How about wrapping switch-case into `classnames(switch() { case '': ... }`?

Comment (upacyxou) [reply src/components/ui/Button.tsx:106]: it doesn't work, I tried

File: src/icons/Loading.tsx [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+const icon = (small?: boolean, screenCentre?: boolean) =>
+  classnames(
+    animation('animate-spin'),
+    textColor('text-inherit'),
+    width(small ? 'w-3' : 'w-5'),
+    height(small ? 'h-3' : 'h-5'),
+    inset({ 'top-1/2': screenCentre, 'left-1/2': screenCentre }),
+    position({ absolute: screenCentre })
+  )
+const iconPath = opacity('opacity-100')
```
Comment (T-Damer) [inline src/icons/Loading.tsx]: Btw, why do we need it, I can't really remember 🤔

**Review by T-Damer — APPROVED** [review]

**Pull request merged**